pkg/kubelet/nodeshutdown/systemd/inhibit_linux.go

```diff
@@ -0,0 +1,183 @@
+	}
+
+	return InhibitLock(fd), nil
+}
+
+// ReleaseInhibitLock will release the underlying inhibit lock which will cause the shutdown to start.
+func (bus *DBusCon) ReleaseInhibitLock(lock InhibitLock) error {
+	err := syscall.Close(int(lock))
+
+	if err != nil {
+		return fmt.Errorf("unable to close systemd inhibitor lock: %v", err)
+	}
+
+	return nil
+}
+
+// ReloadLogindConf uses dbus to send a SIGHUP to the systemd-logind service causing logind to reload it's configuration.
+func (bus *DBusCon) ReloadLogindConf() error {
+	systemdService := "org.freedesktop.systemd1"
+	systemdObject := "/org/freedesktop/systemd1"
+	systemdInterface := "org.freedesktop.systemd1.Manager"
+
+	obj := bus.SystemBus.Object(systemdService, dbus.ObjectPath(systemdObject))
+	unit := "systemd-logind.service"
+	who := "all"
+	var signal int32 = 1 // SIGHUP
+
+	call := obj.Call(systemdInterface+".KillUnit", 0, unit, who, signal)
+	if call.Err != nil {
+		return fmt.Errorf("unable to reload logind conf: %v", call.Err)
+	}
+
+	return nil
+}
+
+// MonitorShutdown detects the a node shutdown by watching for "PrepareForShutdown" logind events.
+// see https://www.freedesktop.org/wiki/Software/systemd/inhibit/ for more details.
+func (bus *DBusCon) MonitorShutdown() (<-chan bool, error) {
+	err := bus.SystemBus.AddMatchSignal(dbus.WithMatchInterface(logindInterface), dbus.WithMatchMember("PrepareForShutdown"), dbus.WithMatchObjectPath("/org/freedesktop/login1"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	busChan := make(chan *dbus.Signal, 1)
+	bus.SystemBus.Signal(busChan)
+
+	shutdownChan := make(chan bool, 1)
+
+	go func() {
+		for {
+			select {
+			case event := <-busChan:
+				if event == nil || len(event.Body) == 0 {
+					klog.Errorf("Failed obtaining shutdown event, PrepareForShutdown event was empty")
+				}
+				shutdownActive, ok := event.Body[0].(bool)
+				if !ok {
+					klog.Errorf("Failed obtaining shutdown event, PrepareForShutdown event was not bool type as expected")
+					return
+				}
+				klog.V(1).Infof("Received node shutdown event; shutdown active: %t", shutdownActive)
+				shutdownChan <- shutdownActive
+			}
+		}
+	}()
+
+	return shutdownChan, nil
+}
+
+const (
+	logindConfigDirectory = "/etc/systemd/logind.conf.d/"
+	kubeletLogindConf     = "kubelet.conf"
+)
+
+// OverrideInhibitDelay writes a config file to logind overriding InhibitDelayMaxSec to the value desired.
+func (bus *DBusCon) OverrideInhibitDelay(inhibitDelayMax time.Duration) error {
+	err := os.MkdirAll(logindConfigDirectory, 0755)
+	if err != nil {
+		return fmt.Errorf("failed creating %v directory: %v", logindConfigDirectory, err)
+	}
+
+	inhibitOverride := fmt.Sprintf(`# Kubelet logind override
+[Login]
+InhibitDelayMaxSec=%.0f
+`, inhibitDelayMax.Seconds())
+
+	logindOverridePath := filepath.Join(logindConfigDirectory, kubeletLogindConf)
+	if err := ioutil.WriteFile(logindOverridePath, []byte(inhibitOverride), 0755); err != nil {
```

Comment:
what if somebody already have this file? Can there be more settings than just InhibitDelayMaxSec?

Reply:
The file written here will be `/etc/systemd/logind.conf.d/99-kubelet.conf`. Since it called `99-kubelet.conf` I would hope that it's reserved only for kubelet to write. 

The idea here is that by default the `InhibitDelayMaxSec` defaults to 5 sec. If the user requested higher shutdown grace period via kubelet, kubelet will attempt to write this config file to override the system's `InhibitDelayMaxSec` setting.

Logind will go to various directories to search for all the logind override config files. It could theoretically happen that some other process on the systemd could write a logind override config file and it would be precedence. From logind docs:

> When multiple files specify the same option, for options which accept just a single value, the entry in the file with the lexicographically latest name takes precedence. 

As a result, this is best effort. If the update was not successful, the shutdown manager will exit. That's the responsibility of the check in `nodeshutdown_manager_linux.go`:

```
// Read the current inhibitDelay again, if the override was successful, currentInhibitDelay will be equal to shutdownGracePeriodRequested.
updatedInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
if err != nil {
    return err
}

if updatedInhibitDelay != m.shutdownGracePeriodRequested {
    return fmt.Errorf("node shutdown manager was unable to update logind InhibitDelayMaxSec to %v (ShutdownGracePeriod), current value of InhibitDelayMaxSec (%v) is less than requested ShutdownGracePeriod", m.shutdownGracePeriodRequested, updatedInhibitDelay)
}
```